src/dataflow/src/server.rs

```diff
@@ -552,7 +552,7 @@ where
                         &new_frontier
                     ));
                     if prev_frontier != &new_frontier {
-                        add_progress(*id, &prev_frontier, &new_frontier, &mut progress);
+                        add_progress(*id, &new_frontier, &prev_frontier, &mut progress);
```

Comment:
Is it possible (and worth it) to add a test that would fail with the wrong order?

Reply:
Potentially yes, but I don't know how viable that is because that code is not very testable and/or this is about the system as a whole. We would observe this in timestamps not being compacted in the SQLite database. @ruchirK Are there any tests that check what is going on in that database?

Reply:
So -- the coordinator tracks a `MutableAntichain` that keeps the `upper` frontier for each source / index. Logically, the `MutableAntichain` kind of reference counts times that are being used, so for example if you run materialize with 5 workers and the mutable antichain contains something like

```
(X, 2)
(Y, 3)
```

then that indicates that 2 workers have reported an upper = `X` and 3 workers have reported an upper = `Y` and the overall frontier is `X`

in this case the antichain would be filled with things like

```
(0, 5)
(X, -3)
(Y, -2)
```

and maybe theres something we could assert here that this mutable antichain can only have `n_worker` nonzero elements. @mjibson might have more opinions

Reply:
For testing timestamp compaction specifically, you can do something like:

* create an unmaterialized source
* create a materialized view based on that source
* disable compaction on that view's index (so that only the source has compaction)
* figure out a time that you can read from
* drop the materialized view
* wait `compaction-window` seconds
* create a new materialized view
* try to select from the old time

(this is a long process but also the shortest thing I could think of to directly exercise disk compaction)

Reply:
I'll try to write a quick test that asserts that the `upper` frontiers never include a negative entry later today.

@mjibson and I had a talk where we concluded that likely the best way forward with testing this stuff is if we expose the timestamp binding info as some kind of system table, and then query that in tests

Reply:
I added the test to check for non-negative `upper` here https://github.com/MaterializeInc/materialize/pull/7261